Problem: no compact storage for archive node (#791)
crypto-org-chain/cronos: versiondb/memiavl/README.md

- Historical files can be compressed with zlib, because it don't need to support random access.

### IAVL Snapshot

IAVL snapshot is composed by four files:

- `metadata`, two integers:

  ```
  version: int64
  root node index: int64
  ```

- `nodes`, array of fixed size(64bytes) nodes, the node format is like this:

  ```
  height  : int8          // padded to 4bytes
  version : int32
  size    : int64
  key     : int64         // offset in keys file
  left    : int32         // inner node only
  right   : int32         // inner node only
  value   : int64 offset  // offset in values file, leaf node only
  hash    : [32]byte
  ```
  The node has fixed length, can be indexed directly. The nodes reference each other with the index, nodes are written in post-order, so the root node is always placed at the end.

  Some integers are using `int32`, should be enough in forseeable future, but could be changed to `int64` to be safer.

  The implementation will read the mmap-ed content in a zero-copy way, don't use extra node cache, just rely on OS page cache.

- `keys`, sequence of length prefixed node keys, deduplicated.

  ```
  size: int16
  payload
  ...
  ```

>   Key size is encoded in `int16`, so maximum key length supported is `65536`. 

Reviewer: should be `2**15 - 1` if it's `int16`?

- `values`, sequence of length prefixed leaf node values.

  ```
  size: int32
  payload
  ...
  ```

>   Value size is encoded in `int32`, so maximum key length supported is `2**32`, around 4G.

Reviewer: should be `2**31 - 1` because not `uint32`?

Author: changed to `uint32`.

- Better read amplification, the IAVL snapshot is a plain file, the nodes are referenced with offset, the read amplification is simply 1.
- Better space amplification, the archived change sets are much more compact than current IAVL tree, in our test case, the ratio could be as large as 1:100. We don't need to keep too old IAVL snapshots, because versiondb will handle the historical key-value queries, IAVL tree only takes care of merkle proof generations for blocks within an unbonding period. In very rare cases that do need IAVL tree of very old version, you can always replay the change sets from the genesis.

## File Formats

> NOTICE: the integers are always encoded with little endianness.

### Change Set File

```
version: int64
size: int64         // size of whole payload
payload:
  delete: int8
  keyLen: varint-uint64
  key
  [                 // if delete is false
    valueLen: varint-uint64
    value
  ]

repeat with next version
```

- Change set files can be splited with certain block ranges for incremental backup and restoration.

> - Historical files can be compressed with zlib, because it don't need to support random access.

Reviewer 2: ```suggestion
- Historical files can be compressed with zlib, because it doesn't need to support random access.
```

### IAVL Snapshot

IAVL snapshot is composed by four files:

- `metadata`, 16bytes:

  ```
  magic: uint32
  root node index: uint32
  version: uint64
  ```

- `nodes`, array of fixed size(64bytes) nodes, the node format is like this:

  ```
  height  : uint8          // padded to 4bytes
  version : uint32
  size    : uint64
  key     : uint64         // offset in keys file
  left    : uint32         // inner node only
  right   : uint32         // inner node only
  value   : uint64 offset  // offset in values file, leaf node only
  hash    : [32]byte
  ```
  The node has fixed length, can be indexed directly. The nodes reference each other with the index, nodes are written in post-order, so the root node is always placed at the end.

  Some integers are using `uint32`, should be enough in forseeable future, but could be changed to `uint64` to be safer.

>   The implementation will read the mmap-ed content in a zero-copy way, don't use extra node cache, just rely on OS page cache.

Reviewer 2: ```suggestion
  The implementation will read the mmap-ed content in a zero-copy way, won't use extra node cache, it will only rely on the OS page cache.
```

- `keys`, sequence of length prefixed leaf node keys, ordered and no duplication.

  ```
  size: uint32
  payload
  *repeat*
  ```

  Key size is encoded in `uint32`, so the maximum key length supported is `1<<32-1`, around 4G.

- `values`, sequence of length prefixed leaf node values.

  ```
  size: uint32
  payload
  *repeat*
  ```

  Value size is encoded in `uint32`, so maximum value length supported is `1<<32-1`, around 4G.

#### Compression

The items in snapshot reference with each other by file offsets, we can apply some block compression techniques to compress keys and values files while maintain random accessbility by uncompressed file offset, for example zstd's experimental seekable format[^1].

### VersionDB

> [VersionDB](../README.md) is to support query and iterating historical versions of key-values pairs, currently implemented with rocksdb's experimental user-defined timestamp feature, support query and iterate key-value pairs by version, it's an alternative way to support grpc query service, and much more compact than IAVL tree, similar in size with the compressed change set files.

Reviewer 2: ```suggestion
[VersionDB](../README.md) is to support query and iterating historical versions of key-values pairs, currently implemented with rocksdb's experimental user-defined timestamp feature, support query and iterate key-value pairs by version, it's an alternative way to support grpc query service, and much more compact than IAVL trees, similar in size with the compressed change set files.
```
